Add graceful close

cyx opened the pull request.

cyx committed:
Add graceful close

apg commented on librato.go:
```diff
@@ -29,29 +29,54 @@ type Reporter struct {
+		case now := <-ticker:
+			var metrics Batch
+			var err error
+			if metrics, err = self.BuildRequest(now, self.Registry); err != nil {
```
Where is `now` ?

apg commented on librato.go:
```diff
@@ -29,29 +29,54 @@ type Reporter struct {
+			log.Printf("go-metrics-librato: close received, cleaned up in %dms", time.Since(now)/time.Millisecond)
+			return
 		}
 	}
```
For the love of all things holy, refactor this into a func.

Can easily return `err` and break in the first case, and return `err` and return in the second.

cyx replied:
Yes yes. Refactored. I was under time pressure ^_^

On Tue, Mar 15, 2016 at 3:18 PM Andrew Gwozdziewycz <
notifications@github.com> wrote:

> In librato.go
> https://github.com/heroku/go-metrics-librato/pull/1#discussion_r56251037
> :
> 
> > -           }
> >   +
> > -       case <-self.closer:
> > -           now := time.Now()
> > -           var metrics Batch
> > -           var err error
> > -           if metrics, err = self.BuildRequest(now, self.Registry); err != nil {
> > -               log.Printf("ERROR constructing librato request body %s", err)
> > -               return
> > -           }
> > -           if err := metricsApi.PostMetrics(metrics); err != nil {
> > -               log.Printf("ERROR sending metrics to librato %s", err)
> > -               return
> > -           }
> > -           log.Printf("go-metrics-librato: close received, cleaned up in %dms", time.Since(now)/time.Millisecond)
> > -           return
> >       }
> >   }
> 
> For the love of all things holy, refactor this into a func.
> 
> Can easily return err and break in the first case, and return err and
> return in the second.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/heroku/go-metrics-librato/pull/1/files/fa290d95342bf6e1ae90461fa72c3734e6bc92c4#r56251037

cyx committed:
Use context for librato reporter

@apg goes without saying that having this merged upstream will
be less likely -- but I like the new interface.

cyx committed:
Refactor the post function

apg commented on librato.go:
```diff
@@ -29,33 +31,48 @@ type Reporter struct {
-func Librato(r metrics.Registry, d time.Duration, e string, t string, s string, p []float64, u time.Duration) {
-	NewReporter(r, d, e, t, s, p, u).Run()
+func Librato(ctx context.Context, r metrics.Registry, d time.Duration, e string, t string, s string, p []float64, u time.Duration) {
+	NewReporter(ctx, r, d, e, t, s, p, u).Run()
```
Do we need to store anything? Just make `Run()` take a context. Then, Librato(ctx, ...), since it calls run, and NewReporter() stays the same. Can then `case <-ctx.Done():` instead.

Commit:
Simple cleanups and a NOTE about intention.

cyx commented:
Good call.
